golang/golang.cpp

```diff
@@ -0,0 +1,124 @@
+                          int capacity, int connectivity, int expansion_add, int expansion_search) {
+    try {
+        config_t config;
+        config.expansion_add = static_cast<std::size_t>(expansion_add);
+        config.expansion_search = static_cast<std::size_t>(expansion_search);
+        config.connectivity = static_cast<std::size_t>(connectivity);
+        config.max_elements = static_cast<std::size_t>(capacity);
+        // todo:: what should those be?
+        config.max_threads_add = std::thread::hardware_concurrency();
+        config.max_threads_search = std::thread::hardware_concurrency();
+
+        accuracy_t accuracy = accuracy_from_name(accuracy_str, accuracy_len);
+        native_index_t index = index_from_name<native_index_t>( //
+            metric_str, metric_len, static_cast<std::size_t>(dimensions), accuracy, config);
+
+        native_index_t* result_ptr = new native_index_t(std::move(index));
+        return result_ptr;
+    } catch (std::exception& e) {
+        printf("error %s\n", e.what());
+    }
+}
+
+void destroy(native_index_t* index) { delete index; }
+
+const char* save(native_index_t* index, char* path) {
+    try {
+        index->save(path);
+    } catch (std::exception& e) {
+        return e.what();
+    }
+}
+
+const char* load(native_index_t* index, char* path) {
+    try {
+        index->load(path);
+    } catch (std::exception& e) {
+        return e.what();
+    }
+}
+
+const char* view(native_index_t* index, char* path) {
+    try {
+        index->view(path);
+    } catch (std::exception& e) {
+        return e.what();
+    }
+}
+
+int size(native_index_t* index) { return index->size(); }
+
+int connectivity(native_index_t* index) { return index->connectivity(); }
+
+// q:: why is this not "dimension"?
+int dimensions(native_index_t* index) { return index->dimensions(); }
+int capacity(native_index_t* index) { return index->capacity(); }
+
+const char* set_capacity(native_index_t* index, int capacity) {
+    try {
+        index->reserve(capacity);
+    } catch (std::exception& e) {
+        // todo:: q:: does e need to free'd?
+        return e.what();
+    }
+}
+
+const char* add(native_index_t* index, int label, float* vector) {
+    // q:: I followed the java example to have try catches everywhere
+    // but they are kind of useless as most errors areoutside of cpp so
+    // translate into sefaults and are not caught by the runtime
+    try {
+        // todo:: the code cannot currently be used in a multithreaded environment
+        // because of golang's roaming gorountines, but once we have a way to
+        // work around it, we should revisit this as this check has a toctou race
+        if (index->size() + 1 >= index->capacity())
+            index->reserve(ceil2(index->size() + 1));
+        index->add(label, vector);
+    } catch (std::exception& e) {
+        return e.what();
```

Comment:
This isn't safe, as you are returning a pointer to the string that may already be destroyed.